File: ThriftSharp/Utilities/TaskEx.cs
// Copyright (c) 2014-15 Solal Pirelli
// This code is licensed under the MIT License (see Licence.txt for details).

using System;
using System.Threading;
using System.Threading.Tasks;

namespace ThriftSharp.Utilities
{
    /// <summary>
    /// Task utility methods.
    /// </summary>
    /// <remarks>
    /// The naming of this class is not optimal, but a class named TaskExtensions already exists in S.T.Tasks.
    /// </remarks>
    internal static class TaskEx
    {
        // The following two methods are adapted from http://blogs.msdn.com/b/pfxteam/archive/2011/11/10/10235834.aspx

        /// <summary>
        /// Makes the <see cref="Task{TResult}" /> timeout after the specified amount of time.
        /// </summary>
        public static Task<TResult> TimeoutAfter<TResult>( this Task<TResult> task, TimeSpan timeout )
        {
            if ( task.IsCompleted || timeout == Timeout.InfiniteTimeSpan )
            {
                return task;
            }

            var resultSource = new TaskCompletionSource<TResult>();
            var timeoutSource = new CancellationTokenSource();

            Task.Delay( timeout, timeoutSource.Token )
                .ContinueWith( ( _, state ) => ( (TaskCompletionSource<TResult>) state ).TrySetCanceled(), resultSource );

            task.ContinueWith( ( antecedent, state ) =>
                {
                    var tuple = (Tuple<CancellationTokenSource, TaskCompletionSource<TResult>>) state;
                    MarshalTaskResults( antecedent, tuple.Item2 );
                    tuple.Item1.Cancel();
                },
                Tuple.Create( timeoutSource, resultSource )
            );

            return resultSource.Task;
        }

        /// <summary>
        /// Marshals the results of the specified task to the specified task source.
        /// </summary>
        private static void MarshalTaskResults<TResult>( Task<TResult> source, TaskCompletionSource<TResult> proxy )
        {
            switch ( source.Status )
            {
                case TaskStatus.Faulted:
                    proxy.TrySetException( source.Exception.InnerException );
                    break;

                case TaskStatus.Canceled:
                    proxy.TrySetCanceled();
                    break;

                case TaskStatus.RanToCompletion:
                    proxy.TrySetResult( source.Result );
                    break;
            }
        }
    }
}

File: ThriftSharp.Tests/TaskEx.fs
module ThriftSharp.Tests.``TaskEx: TimeoutAfter``

open System
open System.Threading
open System.Threading.Tasks
open Xunit
open ThriftSharp.Utilities

let MaxTimeout = TimeSpan.FromMilliseconds(40.0)

[<Fact>]
let ``Already-completed task``() = asTask <| async {
    let! result = TaskEx.TimeoutAfter(Task.FromResult(42), MaxTimeout) |> Async.AwaitTask
    result <=> 42
}    

[<Fact>]
let ``Already-faulted task``() = asTask <| async {
    let! exn = Assert.ThrowsAsync<Exception>(fun () -> 
            TaskEx.TimeoutAfter(Task.FromException<int>(Exception("Oops")), MaxTimeout) :> Task) |> Async.AwaitTask
    exn.Message <=> "Oops"
}    

[<Fact>]
let ``Already-canceled task``() = asTask <| async {
    do! Assert.ThrowsAnyAsync<OperationCanceledException>(fun () ->
            TaskEx.TimeoutAfter(Task.FromCanceled<int>(CancellationToken(true)), MaxTimeout) :> Task) |> Async.AwaitTask |> Async.Ignore
}

[<Fact>]
let ``Quick task``() = asTask <| async {
    let source = TaskCompletionSource<int>()
    let task = Task.Delay(10).ContinueWith(fun _ -> source.TrySetResult(42))
    let! result = TaskEx.TimeoutAfter(source.Task, MaxTimeout) |> Async.AwaitTask
    result <=> 42
    do! task |> Async.AwaitTask |> Async.Ignore
}

[<Fact>]
let ``Quickly faulted task``() = asTask <| async {
    let source = TaskCompletionSource<int>()
    let task = Task.Delay(10).ContinueWith(fun _ -> source.TrySetException(Exception("Oops")))
    let! exn = Assert.ThrowsAsync<Exception>(fun () -> TaskEx.TimeoutAfter(source.Task, MaxTimeout) :> Task) |> Async.AwaitTask
    exn.Message <=> "Oops"
    do! task |> Async.AwaitTask |> Async.Ignore
}

[<Fact>]
let ``Quickly canceled task``() = asTask <| async {
    let source = TaskCompletionSource<int>()
    let task = Task.Delay(10).ContinueWith(fun _ -> source.TrySetCanceled(CancellationToken(false)))
    do! Assert.ThrowsAnyAsync<OperationCanceledException>(fun () ->
            TaskEx.TimeoutAfter(source.Task, MaxTimeout) :> Task) |> Async.AwaitTask |> Async.Ignore
    do! task |> Async.AwaitTask |> Async.Ignore
}

[<Fact>]
let ``Never-completing task``() = asTask <| async {
    let source = TaskCompletionSource<int>()
    do! Assert.ThrowsAnyAsync<OperationCanceledException>(fun () ->
            TaskEx.TimeoutAfter(source.Task, MaxTimeout) :> Task) |> Async.AwaitTask |> Async.Ignore
}

[<Fact>]
let ``Very lengthy task``() = asTask <| async {
    let source = TaskCompletionSource<int>()
    Task.Delay(100).ContinueWith(fun _ -> source.TrySetResult(0)) |> ignore
    do! Assert.ThrowsAnyAsync<OperationCanceledException>(fun () ->
            TaskEx.TimeoutAfter(source.Task, MaxTimeout) :> Task) |> Async.AwaitTask |> Async.Ignore
}